### PR TITLE
Add static builds per cxx_version

### LINKED ISSUE #29: Compile (static) abseil for different C++ versions?
Abseil has a bit of a special situation because the API changes with the C++ version (e.g. using `std::string_view` for C++17 and onwards, but `absl::string_view` below that). This can lead to pretty cryptic linker errors for projects consuming the C++ libs of abseil, because the wrong symbols will be present depending on the version that the target project is compiled with.

Considering that conda-forge allows individual projects to set different C++ standards, we should IMO consider using separate outputs depending on the cxx version used to compile (this minimum can be enforced with CMake's `target_compile_features` that's already [used](https://github.com/abseil/abseil-cpp/blob/20211102.0/CMake/AbseilHelpers.cmake#L257-L261) upstream but still off by default).

If I'm not off with this, the newest abseil-cpp builds are already effectively requiring C++17 (since https://github.com/conda-forge/abseil-cpp-feedstock/commit/fef86bfd67fca0ea0833a96b2678cf738f8fc3b3). I ran into this while testing older C++ standards in https://github.com/conda-forge/sentencepiece-feedstock/pull/26, and running into missing symbols on osx/linux where there weren't any when compiling with C++17.

CC @conda-forge/abseil-cpp

### PR DESCRIPTION
**Pull request by h-vetinari**
Differently from almost any other library, abseil's ABI depends on the C++ standard version used to compile it, and abseil specifically only supports compiling all dependencies with a given standard version.

This is obviously infeasible for an ecosystem like conda-forge, where different packages have different needs (some cannot yet be compiled with e.g. C++17, some already require that).

This makes the general case - a dynamic library - infeasible, because that dynamic library cannot coexist in the same environment with separate ABIs. Since abseil 20211102, we are on C++17 for the dynamic builds, and some libraries (like tensorflow) had issues with this. An experiment with switching windows builds to a higher C++ standard also failed, as e.g. arrow still needs C++11.

In order to cater to different needs (while keeping dynamic libs as a default on unix), this PR add static libraries for all available C++ standards. This way, a project can specify the C++ ABI it needs from abseil. This can even be done without opting out of the migrator, e.g. by:
```
requirements:
  host:
    - libabseil_static             # unpinned to catch migrator
    - libabseil_static =*=_cxx11   # but pin build string to e.g. C++11
```

This is compatible with the [CFEP 18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md) caveat for exceptions
> Building static archives (in separated packages) is still useful for some non-common workflows. These don't represent the majority of uses of our packages but can profit a lot from our using our existing infrastructure.

Abseil is such an exception (even though we will continue to prefer the dynamic libs for those who can live with C++17).

Closes #29

### CODE REVIEW

**Comment by h-vetinari** [comment]
@wolfv 
I realise what I'm trying to do here is quite a hack, but it's the only way I can see for building the same output with different build-strings.

The problem is that I want to select one of these as a wrapper package, but while the initial solve works out until
```
BUILD START: ['libabseil_dynamic-20211102.0-cxx17_h48a1fff_2.tar.bz2', 'libabseil-20211102.0-h93e1e8c_2.tar.bz2', 'abseil-cpp-20211102.0-ha770c72_2.tar.bz2']
```
it then immediately fails to find one of those packages.
```
Reloading output folder: /home/conda/feedstock_root/build_artifacts
Mamba failed to solve:
 - libabseil 20211102.0.* h93e1e8c_2

with channels:

The reported errors are:
- Encountered problems while solving:
-   - nothing provides requested libabseil 20211102.0.* h93e1e8c_2
```

Any insight into where mamba is failing here?

**Comment by h-vetinari** [comment]
@conda-forge/core, I'd like to have some more eyes on this to have people agree to this exception to CFEP-18:

More details in #29, but in short:
* the ABI of abseil depends on the C++ standard version used to compile (and upstream only supports the scenario where an entire stack is compiled against the same version).
* we cannot enforce a uniform C++ standard across conda-forge; some packages still need C++11, some already need C++17.
* it's therefore impossible to do _everything_ with a shared lib, because we cannot have the same version of abseil several times (per respective C++ std ABI) as a runtime dependence.

This PR proposes to :
* stay with shared libs on unix, compiled with C++17 (upstream doesn't support the full library on windows as a shared build).
* keep using the shared lib for all feedstock capable of using C++17.
* provide an escape hatch in the form of static libs that feedstocks stuck on older C++ standards can use to compile against.

I'm also using the chance of such an overhaul to align this feedstock with the discussion from https://github.com/conda-forge/conda-forge.github.io/issues/1073. Happy to take that out if people are opposed to that.

**Comment by h-vetinari** [comment]
To underscore the need for this: current gRPC versions need C++14 and fail in CI when hacked to use C++11. OTOH arrow still needs C++11 on windows. This provides a way out for feedstocks to progress at their respective pace.

To demonstrate my point, I've pushed the current state of this PR into the `abseil_dev` label (modulo a rename `_dynamic` -> `_shared` that's missing in the dev branch) and it passes the gRPC CI in https://github.com/conda-forge/grpc-cpp-feedstock/pull/196.

File: recipe/meta.yaml [hunk 1]
```diff
@@ -14,42 +25,95 @@ source:
...
-    "symbolize", "synchronization", "time", "time_zone"
-] %}
-test:
-  commands:
-    # windows (DLL + import library)
-    - if not exist %LIBRARY_BIN%\\abseil_dll.dll exit 1  # [win]
-    - if not exist %LIBRARY_LIB%\\abseil_dll.lib exit 1  # [win]
+  {% else %}  # shared_libs == "OFF"
+
+  - name: libabseil_static
```
Comment (carterbox) [inline recipe/meta.yaml:55]:
```suggestion
  - name: libabseil-static
```
package names should use "-" instead of "_".

https://github.com/conda-forge/cfep/blob/main/cfep-18.md

**Review by carterbox — CHANGES_REQUESTED** [review]

File: recipe/meta.yaml [hunk 2]
```diff
@@ -14,42 +25,95 @@ source:
...
 
-requirements:
-  build:
-    - {{ compiler('c') }}
-    - {{ compiler('cxx') }}
-    - cmake
-    - ninja
+outputs:
+  {% if shared_libs == "ON" %}
+  - name: libabseil_shared
```
Comment (jaimergp) [inline recipe/meta.yaml:32]: Do we need this output name? I would assume that `libabseil` (the alias below) is shared.

Comment (h-vetinari) [reply recipe/meta.yaml:32]: I've tried to keep the package doing what it used to be doing, i.e. shared on unix, static on windows. That's just to ease transition from the current state of affairs though.

I'm happy to get rid of one of that though, would make it clearer that windows has no shared libs, for example (and we could still keep the current behaviour in `abseil-cpp`).

Comment (h-vetinari) [reply recipe/meta.yaml]: This is done now

**Comment by h-vetinari** [comment]
## TL;DR: Planning to merge this next Monday EOD AOE

@conda-forge/core 
This was discussed in the dev meeting last week ([minutes](https://github.com/conda-forge/conda-forge.github.io/blob/main/src/orga/minutes/2022-07-27.md)). I had suggested to leave 1-2 weeks for comment, and now a bit over one week has passed.

I've got the reinstated migrator [ready](https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/3181) to go as well (at least I think so; happy for review).

Please comment if you have any remaining questions or objections.

PS. One thought I kept thinking about was that despite fully shared builds not being [supported](https://github.com/abseil/abseil-cpp/pull/1115#discussion_r838709189) upstream, it would still be good to keep building a shared(-as-much-as-possible) build on windows. I've added that in the last two commits.
